core/src/main/java/com/amplifyframework/analytics/BasicAnalyticsEvent.java

```diff
@@ -15,31 +15,148 @@
 
 package com.amplifyframework.analytics;
 
+import androidx.annotation.NonNull;
+
 /**
- * Represents event in its most general form and can have different properties.
+ * BasicAnalyticsEvent is a custom analytics event that holds a type and a number of
+ * {@link Properties}. This data object is used to indicate an event occurred such as a user taking
+ * an action in your application.
+ *
+ * <pre>
+ *     BasicAnalyticsEvent event = BasicAnalyticsEvent.builder()
+ *          .type("LikedPost")
+ *          .addProperty("PostType", "UserImage")
+ *          .addProperty("LikedUserID", 78219)
+ *          .addProperty("FirstLike", true)
+ *          .build();
+ * </pre>
+ *
+ * Once built, a BasicAnalyticsEvent can be submitted to an analytics plugin through
+ * {@link AnalyticsCategory#recordEvent(AnalyticsEvent)}.
  */
 public final class BasicAnalyticsEvent implements AnalyticsEvent {
-    private final String name;
+    private final String type;
     private final Properties properties;
 
-    /**
-     * Construct a general analytics event.
-     *
-     * @param name name for the event.
-     * @param properties event properties.
-     */
-    public BasicAnalyticsEvent(String name, Properties properties) {
-        this.name = name;
+    private BasicAnalyticsEvent(String type, Properties properties) {
+        this.type = type;
         this.properties = properties;
     }
 
+    /**
+     * Returns the type of the event.
+     *
+     * @return The type of the event
+     */
     @Override
-    public String getName() {
-        return name;
+    @NonNull
+    public String getType() {
+        return type;
     }
 
+    /**
+     * Returns the {@link Properties} of the event.
+     *
+     * @return The {@link Properties} of the event
+     */
     @Override
+    @NonNull
     public Properties getProperties() {
         return properties;
     }
+
+    /**
+     * Returns a new {@link Builder} to configure an instance of BasicAnalyticsEvent.
+     *
+     * @return a {@link Builder}
+     */
+    @NonNull
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Builder is used to create and configure an instance of {@link BasicAnalyticsEvent}. Its
+     * methods return the Builder instance to allow for fluent method chaining. This Builder reuses
+     * {@link Properties.Builder} to construct the properties to store in the event.
+     *
+     * @see Properties
+     */
+    public static final class Builder {
```

Comment:
All of the methods in this builder can also be marked as `@NonNull`, I believe. (A human wouldn't benefit from the cue, on this file. But, a lint-ing program / Kotlin type interpreter, could.)

For the values that we pass into the builder - should we allow them to be null, or no? If no, we have a convention of using `Objects.requireNonNull(...)` to "fail fast" as soon as we see a bad user-provided value, at its first location of use.

Reply:
Ahh, yes, you are correct. Thanks! I'll get the hang of this eventually. Adding the fail-fast incantation where appropriate.